### ISSUE TITLE
Inhomogeneous Dirichlet and Neumann conditions in parallel for Example 16p heat equation

### ISSUE THREAD

**Issue by robert-chiodi** [mfem/mfem#1811]
Hello,

I am trying to incorporate inhomogeneous Dirichlet and Neumann conditions into the heat equation code of example 16. Right now, I am just focusing on explicit time integration and am therefore just modifying the `Mult()` function. I have managed to get it working in serial, however, I have yet to get both to work in parallel. I am hoping someone can help me clear up the confusion I have about correctly implementing the conditions in parallel.

To implement the Dirichlet conditions, I followed issue #1720 and use `K->FormLinearSystem(ess_tdof_list,x,b,X,B)` to set the Dirichlet conditions of the right hand side. The documentation for `ParBilinearForm::FormLinearSystem` states that `x` should be of GridFunction size and `b` should be of LinearForm size. I believe `u` enters `ConductionOperator::Mult` containing only true DOFs, so I have created `ParGridFunction` objects for `u`, and `z` (named `tmp_u` and `tmp_z`), which I set via ParGridFunction::SetFromTrueDofs(). This seems to work for both serial and parallel if no inhomogeneous Neumann condition is used, with the solution recovered via `M_solver.Mult(B,du_dt)`. 

For the inhomogeneous Neumann condition, I have used a LinearForm with` AddBoundaryIntegrator(new mfem::BoundaryLFIntegrator(ten), ess_bdr)`, which I then assemble and add to the `tmp_z` ParGridFunction object. I then use `M->FormLinearSystem` with a `tmp_du_dt` ParGridFunction object obtained via `SetFromTrueDofs(du_dt)`. The solution for `tmp_du_dt` is then recovered using `M_solver` and cast back to `du_dt` via `tmp_du_dt.GetTrueDofs(du_dt)`. This works in serial (with mpiexec -np 1 used), but blows up (throwing the assertion `Assertion failed: (IsFinite(nom)) is false:` in CGSolver::Mult`) in different amounts of iterations depending on the number of processes used.

My main questions are:

- Was I correct in using the `tmp_u`, `tmp_z`, `tmp_du_dt` ParGridFunctions to get it working in parallel or is there a better approach?
- Is adding `tmp_z` and the LinearForm holding the Neumann conditions valid?
- Is there an obvious error in my approach for inhomogeneous Dirichlet/Neumann conditions?

I am open to any/all suggestions on how to best achieve setting inhomogeneous Dirichlet and Neumann conditions. I have also included the code below as it might be easier to understand that way.

Thank you in advance for any help or guidance you can give.

----------------------------------------------------------------------

```
void ConductionOperator::Mult(const mfem::Vector& u,
                              mfem::Vector& du_dt) const {
  // Compute:
  //    du_dt = M^{-1}*-K(u)
  // for du_dt
  K_op->Mult(u, z);
  z.Neg();  // z = -z

  // Elevate true DOF vector u to ParGridFunction
  mfem::ParGridFunction tmp_u(&fespace);
  tmp_u.SetFromTrueDofs(u);

  // Elevate true DOF vector z to ParGridFunction
  mfem::ParGridFunction tmp_z(&fespace);
  tmp_z.SetFromTrueDofs(z);

  // Example inhomogeneous Neumann condition.
  mfem::ParLinearForm neumann(&fespace);
  mfem::ConstantCoefficient ten(10.0);

  // Set Neumann on top/bottom BC's of square mesh.
  mfem::Array<int> ess_bdr(4);
  ess_bdr[2] = 1;
  ess_bdr[3] = 1;
  neumann.AddBoundaryIntegrator(new mfem::BoundaryLFIntegrator(ten), ess_bdr);
  neumann.Assemble();

  // Incorporate Neumann into RHS.
  tmp_z.Add(1.0, neumann);

  // Form system, accounting for any Dirichlet conditions
  // stored in u (set in a separate part of the code).
  K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);

  // Elevate true DOF vector du_dt to ParGridFunction
  mfem::ParGridFunction tmp_du_dt(&fespace);
  tmp_du_dt.SetFromTrueDofs(du_dt);

  // Form system to solve
  M->FormLinearSystem(ess_tdof_list, tmp_du_dt, tmp_z, A, X, B);

  // M hasn't changed since construction, has same ess_tdof_list
  // when constructed and used to set M_solver's operator.
  M_solver.Mult(B, X);
  M->RecoverFEMSolution(X, tmp_z, tmp_du_dt);
  tmp_du_dt.GetTrueDofs(du_dt);
}
```

**Comment by chang-yiyi** [comment]
Hi @robert-chiodi ,

Why would you need `K->FormLinearSystem()`, i.e. the below
```
  // Form system, accounting for any Dirichlet conditions
  // stored in u (set in a separate part of the code).
  K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);
```
I think you maybe need to set `Dirichlet` boundary conditions for `du_dt`, not `u`.

**Comment by robert-chiodi** [comment]
Hi @chang-yiyi ,

Thank you for your response. I believe those lines,
```
  // Form system, accounting for any Dirichlet conditions
  // stored in u (set in a separate part of the code).
  K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);
```
need to be set in order to incorporate the effect of the essential boundary conditions on the RHS `B`. At least, that is what I believe the result of discussion in #1720 reached. That part seemed to be working correctly in serial and parallel, with the inhomogeneous Neumann condition leading to most of the troubles. I am very new to MFEM though and there could certainly be a better, more correct, or more idiomatic way to incorporate the essential boundary conditions.

**Comment by jandrej** [comment]
1) Yes, you are right on the point that the parameters in `::Mult` are T-vectors (true dof size).

2) Repeatedly calling `FormLinearSystem` should be OK, since it "remembers" the essential dofs.

Do you apply the exact same boundary attribute array
```
  mfem::Array<int> ess_bdr(4);
  ess_bdr[2] = 1;
  ess_bdr[3] = 1;
```
in the `main` function? You call them `ess_bdr` here but use them for Neumann-type BCs.

Adding the Neumann term to `z` should also be fine.

The part that seems off is here
```
// Form system, accounting for any Dirichlet conditions
  // stored in u (set in a separate part of the code).
  K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);

  // Elevate true DOF vector du_dt to ParGridFunction
  mfem::ParGridFunction tmp_du_dt(&fespace);
  tmp_du_dt.SetFromTrueDofs(du_dt);

  // Form system to solve
  M->FormLinearSystem(ess_tdof_list, tmp_du_dt, tmp_z, A, X, B);
```
You shouldn't have to use M for forming the linear system.

**Comment by chang-yiyi** [comment]
Hi,

I want to give some my opinions. If I'm wrong, please correct me.

1. Here the unknown variable is `du_dt`, so the `Dirichlet` and `Neumann` boundary conditions should be set for `du_dt`, instead of `u`.

2. Assume `i` for interior DOFs and `b` for boundary DOFs, set
   ```
   A = | A_ii A_ib |
       | A_bi A_bb |
   ```

   The line `K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B)` will change `B` into 
   ```
   | tmp_z_i - A_ib * tmp_u_b |
   | tmp_z_b                  |
   ``` 
   Here `A` is formed form `K`, and `tmp_z_b = tmp_u_b`.
   And then the line `M->FormLinearSystem(ess_tdof_list, tmp_du_dt, tmp_z, A, X, B)` will change `B` into
   ```
   | tmp_z_i - A_ib * tmp_du_dt_b |
   | tmp_z_b                      |
   ```
   Here `A` is formed form `M`, and `tm_z_b = tmp_du_dt_b`.
   3. Could you please provide a minimal working example?

**Comment by robert-chiodi** [comment]
Hello,

Thank you @jandrej and @chang-yiyi for helping me. I respond to you 

- The use of ess_bdr for the Neumann condition in my example was just a confusing variable name, my apologies. It was not set to be essential in main.cpp, only the other two boundaries were.

-  I am confused about the lack-of-need to use 
`K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B)`, as, without it, it seems like the inhomogeneous Dirichlet BC does not affect the domain.

- I have created a minimal working example by modifying example 16p and included it below. I marked all changes to the code with banners of `// =======...==//`. The main modifications are assumed use of the mesh `inline-quad.mesh` (with a location of  `../data/inline-quad.mesh`), the temperature field `u` initialized to zero, and inhomogeneous Dirichlet conditions of value 10.0 on the top/bottom walls and inhomogeneous Neumann conditions on the left/right walls. I then run the example with either:
-- *Explicit:* mpirun -np 1 example -visit -rs 0 -rp 0 -s 11 -dt 0.001 -o 1
-- *Implicit:* mpirun -np 1 example -visit -rs 0 -rp 0 -s 3 -dt 0.001 -o 1

Both implicit and explicit produce essentially the same results, as expected. The main problem I am having (in both explicit and implicit) is a difference in solution when running in serial (mpirun -np 1) and parallel (mpirun -np 4), and I am not sure what could be causing it. The difference is quite stark, however, shown below from VisIt visualization of the temperature field when using a single mpi process (top picture) vs. four (bottom picture).

Thank you both in advance for any help you may offer.

![serial](https://user-images.githubusercontent.com/47005129/95895504-fdfd3100-0d47-11eb-8c5b-27d3ed956974.png)
![parallel_4_core](https://user-images.githubusercontent.com/47005129/95895519-05bcd580-0d48-11eb-9a11-38cd731029f5.png)

```
//                       MFEM Example 16 - Parallel Version
//
// Compile with: make ex16p
//
// Sample runs:  mpirun -np 4 ex16p
//               mpirun -np 4 ex16p -m ../data/inline-tri.mesh
//               mpirun -np 4 ex16p -m ../data/disc-nurbs.mesh -tf 2
//               mpirun -np 4 ex16p -s 1 -a 0.0 -k 1.0
//               mpirun -np 4 ex16p -s 2 -a 1.0 -k 0.0
//               mpirun -np 8 ex16p -s 3 -a 0.5 -k 0.5 -o 4
//               mpirun -np 4 ex16p -s 14 -dt 1.0e-4 -tf 4.0e-2 -vs 40
//               mpirun -np 16 ex16p -m ../data/fichera-q2.mesh
//               mpirun -np 16 ex16p -m ../data/fichera-mixed.mesh
//               mpirun -np 16 ex16p -m ../data/escher-p2.mesh
//               mpirun -np 8 ex16p -m ../data/beam-tet.mesh -tf 10 -dt 0.1
//               mpirun -np 4 ex16p -m ../data/amr-quad.mesh -o 4 -rs 0 -rp 0
//               mpirun -np 4 ex16p -m ../data/amr-hex.mesh -o 2 -rs 0 -rp 0
//
// Description:  This example solves a time dependent nonlinear heat equation
//               problem of the form du/dt = C(u), with a non-linear diffusion
//               operator C(u) = \nabla \cdot (\kappa + \alpha u) \nabla u.
//
//               The example demonstrates the use of nonlinear operators (the
//               class ConductionOperator defining C(u)), as well as their
//               implicit time integration. Note that implementing the method
//               ConductionOperator::ImplicitSolve is the only requirement for
//               high-order implicit (SDIRK) time integration. Optional saving
//               with ADIOS2 (adios2.readthedocs.io) is also illustrated.
//
//               We recommend viewing examples 2, 9 and 10 before viewing this
//               example.

#include <fstream>
#include <iostream>
#include "mfem.hpp"

using namespace std;
using namespace mfem;

/** After spatial discretization, the conduction model can be written as:
 *
 *     du/dt = M^{-1}(-Ku)
 *
 *  where u is the vector representing the temperature, M is the mass matrix,
 *  and K is the diffusion operator with diffusivity depending on u:
 *  (\kappa + \alpha u).
 *
 *  Class ConductionOperator represents the right-hand side of the above ODE.
 */
class ConductionOperator : public TimeDependentOperator {
 protected:
  ParFiniteElementSpace& fespace;
  Array<int> ess_tdof_list;  // this list remains empty for pure Neumann b.c.

  ParBilinearForm* M;
  ParBilinearForm* K;

  HypreParMatrix Mmat;
  HypreParMatrix Kmat;
  HypreParMatrix* T;  // T = M + dt K
  double current_dt;

  CGSolver M_solver;     // Krylov solver for inverting the mass matrix M
  HypreSmoother M_prec;  // Preconditioner for the mass matrix M

  CGSolver T_solver;     // Implicit solver for T = M + dt K
  HypreSmoother T_prec;  // Preconditioner for the implicit solver

  double alpha, kappa;

  mutable Vector z;  // auxiliary vector

 public:
  ConductionOperator(ParFiniteElementSpace& f, double alpha, double kappa,
                     const Vector& u, const Array<int>& ess_tdof);

  virtual void Mult(const Vector& u, Vector& du_dt) const;
  /** Solve the Backward-Euler equation: k = f(u + dt*k, t), for the unknown k.
      This is the only requirement for high-order SDIRK implicit integration.*/
  virtual void ImplicitSolve(const double dt, const Vector& u, Vector& k);

  /// Update the diffusion BilinearForm K using the given true-dof vector `u`.
  void SetParameters(const Vector& u);

  virtual ~ConductionOperator();
};

double InitialTemperature(const Vector& x);

int main(int argc, char* argv[]) {
  // 1. Initialize MPI.
  int num_procs, myid;
  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
  MPI_Comm_rank(MPI_COMM_WORLD, &myid);

  // 2. Parse command-line options.
  int ser_ref_levels = 2;
  int par_ref_levels = 1;
  int order = 2;
  int ode_solver_type = 3;
  double t_final = 0.5;
  double dt = 1.0e-2;
  double alpha = 1.0e-2;
  double kappa = 0.5;
  bool visualization = true;
  bool visit = false;
  int vis_steps = 5;
  bool adios2 = false;

  int precision = 8;
  cout.precision(precision);

  OptionsParser args(argc, argv);
  args.AddOption(&ser_ref_levels, "-rs", "--refine-serial",
                 "Number of times to refine the mesh uniformly in serial.");
  args.AddOption(&par_ref_levels, "-rp", "--refine-parallel",
                 "Number of times to refine the mesh uniformly in parallel.");
  args.AddOption(&order, "-o", "--order",
                 "Order (degree) of the finite elements.");
  args.AddOption(&ode_solver_type, "-s", "--ode-solver",
                 "ODE solver: 1 - Backward Euler, 2 - SDIRK2, 3 - SDIRK3,\n\t"
                 "\t   11 - Forward Euler, 12 - RK2, 13 - RK3 SSP, 14 - RK4.");
  args.AddOption(&t_final, "-tf", "--t-final", "Final time; start time is 0.");
  args.AddOption(&dt, "-dt", "--time-step", "Time step.");
  args.AddOption(&alpha, "-a", "--alpha", "Alpha coefficient.");
  args.AddOption(&kappa, "-k", "--kappa", "Kappa coefficient offset.");
  args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
                 "--no-visualization",
                 "Enable or disable GLVis visualization.");
  args.AddOption(&visit, "-visit", "--visit-datafiles", "-no-visit",
                 "--no-visit-datafiles",
                 "Save data files for VisIt (visit.l1lnl.gov) visualization.");
  args.AddOption(&vis_steps, "-vs", "--visualization-steps",
                 "Visualize every n-th timestep.");
  args.AddOption(&adios2, "-adios2", "--adios2-streams", "-no-adios2",
                 "--no-adios2-streams", "Save data using adios2 streams.");
  args.Parse();
  if (!args.Good()) {
    args.PrintUsage(cout);
    MPI_Finalize();
    return 1;
  }

  if (myid == 0) {
    args.PrintOptions(cout);
  }

  // ================================================================= //
  // 3. Use inline-quad.mesh file for example.
  // Boundary elements are tagged as:
  // Bottom = 1
  // Right = 2
  // Top = 3
  // Left = 4
  const char* mesh_file = "../data/inline-quad.mesh";
  auto mesh = new Mesh(mesh_file, 1, 1);
  int dim = mesh->Dimension();
  // ================================================================= //

  // 4. Define the ODE solver used for time integration. Several implicit
  //    singly diagonal implicit Runge-Kutta (SDIRK) methods, as well as
  //    explicit Runge-Kutta methods are available.
  ODESolver* ode_solver;
  switch (ode_solver_type) {
    // Implicit L-stable methods
    case 1:
      ode_solver = new BackwardEulerSolver;
      break;
    case 2:
      ode_solver = new SDIRK23Solver(2);
      break;
    case 3:
      ode_solver = new SDIRK33Solver;
      break;
    // Explicit methods
    case 11:
      ode_solver = new ForwardEulerSolver;
      break;
    case 12:
      ode_solver = new RK2Solver(0.5);
      break;  // midpoint method
    case 13:
      ode_solver = new RK3SSPSolver;
      break;
    case 14:
      ode_solver = new RK4Solver;
      break;
    case 15:
      ode_solver = new GeneralizedAlphaSolver(0.5);
      break;
    // Implicit A-stable methods (not L-stable)
    case 22:
      ode_solver = new ImplicitMidpointSolver;
      break;
    case 23:
      ode_solver = new SDIRK23Solver;
      break;
    case 24:
      ode_solver = new SDIRK34Solver;
      break;
    default:
      cout << "Unknown ODE solver type: " << ode_solver_type << '\n';
      delete mesh;
      return 3;
  }

  // 5. Refine the mesh in serial to increase the resolution. In this example
  //    we do 'ser_ref_levels' of uniform refinement, where 'ser_ref_levels' is
  //    a command-line parameter.
  for (int lev = 0; lev < ser_ref_levels; lev++) {
    mesh->UniformRefinement();
  }

  // 6. Define a parallel mesh by a partitioning of the serial mesh. Refine
  //    this mesh further in parallel to increase the resolution. Once the
  //    parallel mesh is defined, the serial mesh can be deleted.
  ParMesh* pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
  delete mesh;
  for (int lev = 0; lev < par_ref_levels; lev++) {
    pmesh->UniformRefinement();
  }

  // 7. Define the vector finite element space representing the current and the
  //    initial temperature, u_ref.
  H1_FECollection fe_coll(order, dim);
  ParFiniteElementSpace fespace(pmesh, &fe_coll);

  int fe_size = fespace.GlobalTrueVSize();
  if (myid == 0) {
    cout << "Number of temperature unknowns: " << fe_size << endl;
  }

  ParGridFunction u_gf(&fespace);

  // ================================================================= //
  // 8. Set u initially to zero.
  ConstantCoefficient zero(0.0);
  u_gf.ProjectCoefficient(zero);
  Vector u;
  u_gf.GetTrueDofs(u);
  // ================================================================= //

  // ================================================================= //
  // 8.b Mark Essential true DOFs for bottom and top walls (will treat
  // as inhomogeneous Dirichlet). Fill boundary of u to be 10.0.
  mfem::Array<int> ess_bdr(pmesh->bdr_attributes.Max());
  mfem::Array<int> ess_tdof_list;
  ess_bdr = 0;
  ess_bdr[0] = 1;  // bottom wall for inline-quad.mesh
  ess_bdr[2] = 1;  // top wall for inline-quad.mesh
  fespace.GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
  ConstantCoefficient ten(10.0);
  u_gf.ProjectBdrCoefficient(ten, ess_bdr);
  u_gf.GetTrueDofs(u);

  // ================================================================= //

  // 9. Initialize the conduction operator and the VisIt visualization.
  ConductionOperator oper(fespace, alpha, kappa, u, ess_tdof_list);

  u_gf.SetFromTrueDofs(u);
  {
    ostringstream mesh_name, sol_name;
    mesh_name << "ex16-mesh." << setfill('0') << setw(6) << myid;
    sol_name << "ex16-init." << setfill('0') << setw(6) << myid;
    ofstream omesh(mesh_name.str().c_str());
    omesh.precision(precision);
    pmesh->Print(omesh);
    ofstream osol(sol_name.str().c_str());
    osol.precision(precision);
    u_gf.Save(osol);
  }

  VisItDataCollection visit_dc("Example16-Parallel", pmesh);
  visit_dc.RegisterField("temperature", &u_gf);
  if (visit) {
    visit_dc.SetCycle(0);
    visit_dc.SetTime(0.0);
    visit_dc.Save();
  }

  // Optionally output a BP (binary pack) file using ADIOS2. This can be
  // visualized with the ParaView VTX reader.
#ifdef MFEM_USE_ADIOS2
  ADIOS2DataCollection* adios2_dc = NULL;
  if (adios2) {
    std::string postfix(mesh_file);
    postfix.erase(0, std::string("../data/").size());
    postfix += "_o" + std::to_string(order);
    postfix += "_solver" + std::to_string(ode_solver_type);
    const std::string collection_name = "ex16-p-" + postfix + ".bp";

    adios2_dc =
        new ADIOS2DataCollection(MPI_COMM_WORLD, collection_name, pmesh);
    adios2_dc->SetParameter("SubStreams", std::to_string(num_procs / 2));
    adios2_dc->RegisterField("temperature", &u_gf);
    adios2_dc->SetCycle(0);
    adios2_dc->SetTime(0.0);
    adios2_dc->Save();
  }
#endif

  socketstream sout;
  if (visualization) {
    char vishost[] = "localhost";
    int visport = 19916;
    sout.open(vishost, visport);
    sout << "parallel " << num_procs << " " << myid << endl;
    int good = sout.good(), all_good;
    MPI_Allreduce(&good, &all_good, 1, MPI_INT, MPI_MIN, pmesh->GetComm());
    if (!all_good) {
      sout.close();
      visualization = false;
      if (myid == 0) {
        cout << "Unable to connect to GLVis server at " << vishost << ':'
             << visport << endl;
        cout << "GLVis visualization disabled.\n";
      }
    } else {
      sout.precision(precision);
      sout << "solution\n" << *pmesh << u_gf;
      sout << "pause\n";
      sout << flush;
      if (myid == 0) {
        cout << "GLVis visualization paused."
             << " Press space (in the GLVis window) to resume it.\n";
      }
    }
  }

  // 10. Perform time-integration (looping over the time iterations, ti, with a
  //     time-step dt).
  ode_solver->Init(oper);
  double t = 0.0;

  bool last_step = false;
  for (int ti = 1; !last_step; ti++) {
    if (t + dt >= t_final - dt / 2) {
      last_step = true;
    }

    ode_solver->Step(u, t, dt);

    if (last_step || (ti % vis_steps) == 0) {
      if (myid == 0) {
        cout << "step " << ti << ", t = " << t << endl;
      }

      u_gf.SetFromTrueDofs(u);
      if (visualization) {
        sout << "parallel " << num_procs << " " << myid << "\n";
        sout << "solution\n" << *pmesh << u_gf << flush;
      }

      if (visit) {
        visit_dc.SetCycle(ti);
        visit_dc.SetTime(t);
        visit_dc.Save();
      }

#ifdef MFEM_USE_ADIOS2
      if (adios2) {
        adios2_dc->SetCycle(ti);
        adios2_dc->SetTime(t);
        adios2_dc->Save();
      }
#endif
    }
    oper.SetParameters(u);
  }

#ifdef MFEM_USE_ADIOS2
  if (adios2) {
    delete adios2_dc;
  }
#endif

  // 11. Save the final solution in parallel. This output can be viewed later
  //     using GLVis: "glvis -np <np> -m ex16-mesh -g ex16-final".
  {
    ostringstream sol_name;
    sol_name << "ex16-final." << setfill('0') << setw(6) << myid;
    ofstream osol(sol_name.str().c_str());
    osol.precision(precision);
    u_gf.Save(osol);
  }

  // 12. Free the used memory.
  delete ode_solver;
  delete pmesh;

  MPI_Finalize();

  return 0;
}

ConductionOperator::ConductionOperator(ParFiniteElementSpace& f, double al,
                                       double kap, const Vector& u,
                                       const Array<int>& ess_tdof)
    : TimeDependentOperator(f.GetTrueVSize(), 0.0),
      fespace(f),
      ess_tdof_list(ess_tdof),
      M(NULL),
      K(NULL),
      T(NULL),
      current_dt(0.0),
      M_solver(f.GetComm()),
      T_solver(f.GetComm()),
      z(height) {
  const double rel_tol = 1e-8;

  M = new ParBilinearForm(&fespace);
  M->AddDomainIntegrator(new MassIntegrator());
  M->Assemble(0);  // keep sparsity pattern of M and K the same
  M->FormSystemMatrix(ess_tdof_list, Mmat);

  M_solver.iterative_mode = false;
  M_solver.SetRelTol(rel_tol);
  M_solver.SetAbsTol(0.0);
  M_solver.SetMaxIter(100);
  M_solver.SetPrintLevel(0);
  M_prec.SetType(HypreSmoother::Jacobi);
  M_solver.SetPreconditioner(M_prec);
  M_solver.SetOperator(Mmat);

  alpha = al;
  kappa = kap;

  T_solver.iterative_mode = false;
  T_solver.SetRelTol(rel_tol);
  T_solver.SetAbsTol(0.0);
  T_solver.SetMaxIter(100);
  T_solver.SetPrintLevel(0);
  T_solver.SetPreconditioner(T_prec);

  SetParameters(u);
}

// // Inhomogeneous Dirichlet on bottom/top walls, homogeneous Neumann on
// // left/right wall.
// // *** Serial and parallel produce very different results ***
// void ConductionOperator::Mult(const Vector& u, Vector& du_dt) const {
//   // Compute:
//   //    du_dt = M^{-1}*-K(u)
//   // for du_dt
//   Kmat.Mult(u, z);
//   z.Neg();  // z = -z

//   ParGridFunction tmp_u(&fespace), tmp_z(&fespace), tmp_du_dt(&fespace);
//   tmp_u.SetFromTrueDofs(u);
//   tmp_z.SetFromTrueDofs(z);
//   tmp_du_dt.SetFromTrueDofs(du_dt);

//   OperatorHandle A;
//   Vector X, B;
//   K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);
//   M_solver.Mult(B, du_dt);
//   du_dt.SetSubVector(ess_tdof_list, 0.0);
// }

// // Inhomogeneous Dirichlet on bottom/top walls, homogeneous Neumann on
// // left/right wall.
// // *** Serial and parallel produce very different results ***
// void ConductionOperator::ImplicitSolve(const double dt, const Vector& u,
//                                        Vector& du_dt) {
//   // Solve the equation:
//   //    du_dt = M^{-1}*[-K(u + dt*du_dt)]
//   // for du_dt
//   if (!T) {
//     T = Add(1.0, Mmat, dt, Kmat);
//     current_dt = dt;
//     T_solver.SetOperator(*T);
//   }
//   MFEM_VERIFY(dt == current_dt, "");  // SDIRK methods use the same dt
//   Kmat.Mult(u, z);
//   z.Neg();

//   ParGridFunction tmp_u(&fespace), tmp_z(&fespace), tmp_du_dt(&fespace);
//   tmp_u.SetFromTrueDofs(u);
//   tmp_z.SetFromTrueDofs(z);
//   tmp_du_dt.SetFromTrueDofs(du_dt);

//   OperatorHandle A;
//   Vector X, B;
//   K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);

//   T_solver.Mult(B, du_dt);
//   du_dt.SetSubVector(ess_tdof_list, 0.0);
// }

// Inhomogeneous Dirichlet on bottom/top walls, inhomogeneous Neumann on
// left/right wall.
// *** Serial and parallel produce very different results ***
void ConductionOperator::Mult(const Vector& u, Vector& du_dt) const {
  // Compute:
  //    du_dt = M^{-1}*-K(u)
  // for du_dt
  Kmat.Mult(u, z);
  z.Neg();  // z = -z

  ParGridFunction tmp_u(&fespace), tmp_z(&fespace), tmp_du_dt(&fespace);
  tmp_u.SetFromTrueDofs(u);
  tmp_z.SetFromTrueDofs(z);
  tmp_du_dt.SetFromTrueDofs(du_dt);

  // Construct and set inhomogeneous Neumann condition
  ParLinearForm neumann(&fespace);
  ConstantCoefficient ten(10.0);
  Array<int> neumann_bdr(fespace.GetParMesh()->bdr_attributes.Max());
  neumann_bdr = 0;
  neumann_bdr[1] = 1;  // Right wall for inline-quad.mesh
  neumann_bdr[3] = 1;  // Left wal for inline-quad.mesh
  neumann.AddBoundaryIntegrator(new mfem::BoundaryLFIntegrator(ten),
                                neumann_bdr);
  neumann.Assemble();

  // Add inhomogeneous Neumann to RHS
  tmp_z.Add(1.0, neumann);

  OperatorHandle A;
  Vector X, B;
  K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);
  M_solver.Mult(B, du_dt);
  du_dt.SetSubVector(ess_tdof_list, 0.0);
}

// Inhomogeneous Dirichlet on bottom/top walls, homogeneous Neumann on
// left/right wall.
// *** Serial and parallel produce very different results ***
void ConductionOperator::ImplicitSolve(const double dt, const Vector& u,
                                       Vector& du_dt) {
  // Solve the equation:
  //    du_dt = M^{-1}*[-K(u + dt*du_dt)]
  // for du_dt
  if (!T) {
    T = Add(1.0, Mmat, dt, Kmat);
    current_dt = dt;
    T_solver.SetOperator(*T);
  }
  MFEM_VERIFY(dt == current_dt, "");  // SDIRK methods use the same dt
  Kmat.Mult(u, z);
  z.Neg();

  ParGridFunction tmp_u(&fespace), tmp_z(&fespace), tmp_du_dt(&fespace);
  tmp_u.SetFromTrueDofs(u);
  tmp_z.SetFromTrueDofs(z);
  tmp_du_dt.SetFromTrueDofs(du_dt);

  // Construct and set inhomogeneous Neumann condition
  ParLinearForm neumann(&fespace);
  ConstantCoefficient ten(10.0);
  Array<int> neumann_bdr(fespace.GetParMesh()->bdr_attributes.Max());
  neumann_bdr = 0;
  neumann_bdr[1] = 1;  // Right wall for inline-quad.mesh
  neumann_bdr[3] = 1;  // Left wal for inline-quad.mesh
  neumann.AddBoundaryIntegrator(new mfem::BoundaryLFIntegrator(ten),
                                neumann_bdr);
  neumann.Assemble();

  // Add inhomogeneous Neumann to RHS
  tmp_z.Add(1.0, neumann);

  OperatorHandle A;
  Vector X, B;
  K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);

  T_solver.Mult(B, du_dt);
  du_dt.SetSubVector(ess_tdof_list, 0.0);
}

void ConductionOperator::SetParameters(const Vector& u) {
  ParGridFunction u_alpha_gf(&fespace);
  u_alpha_gf.SetFromTrueDofs(u);
  for (int i = 0; i < u_alpha_gf.Size(); i++) {
    u_alpha_gf(i) = kappa + alpha * u_alpha_gf(i);
  }

  delete K;
  K = new ParBilinearForm(&fespace);

  GridFunctionCoefficient u_coeff(&u_alpha_gf);

  K->AddDomainIntegrator(new DiffusionIntegrator(u_coeff));
  K->Assemble(0);  // keep sparsity pattern of M and K the same
  K->FormSystemMatrix(ess_tdof_list, Kmat);
  delete T;
  T = NULL;  // re-compute T on the next ImplicitSolve
}

ConductionOperator::~ConductionOperator() {
  delete T;
  delete M;
  delete K;
}

double InitialTemperature(const Vector& x) {
  if (x.Norml2() < 0.5) {
    return 2.0;
  } else {
    return 1.0;
  }
}
```

**Comment by jandrej** [comment]
You apply Dirichlet on the top and bottom with a value of 10.0. I don't know about the behavior of the nonlinearity of the system here, but this Dirichlet boundary should remain the largest value in the system unless there is a reaction in the system.

The plots you show have maximum values of >10.0.

**Comment by robert-chiodi** [comment]
Is that true even in the presence of the inhomogeneous Neumann conditions as well (given a value of 10.0)? I was not surprised by the >10.0 values since for the t = 0.5 of the simulation the left/right boundaries are effectively injecting heat into the system.

If I change the inhomogeneous Neumann to homogeneous Neumann on the left/right walls, the maximum temperature does indeed remain at 10.0 on the top/bottom walls of the domain where the inhomogeneous Dirichlet condition is set (shown in the picture below).

![dirichlet_only](https://user-images.githubusercontent.com/47005129/95903591-96e57980-0d53-11eb-8e84-7492ff9843d5.png)

**Comment by jandrej** [comment]
Got it, missed the part where you put energy into the system.

**Comment by jandrej** [comment]
You can look into `RecoverFEMSolution` and `GetTrueDofs` after the solution of your system. This should take care of proper redistribution of your solution.

**Comment by robert-chiodi** [comment]
Hi @jandrej ,

No worries about missing the energy-injecting part, I appreciate your help. I have reduced the problem to just setting inhomogeneous Dirichlet conditions on the top/bottom and homogeneous Neumann on the left/right (such as in the picture 3 posts above). I then modified the explicit time advancement routine `Mult(...)` in ConductionOperator with `RecoverFEMSolution` and `GetTrueDofs` as 
```
void ConductionOperator::Mult(const Vector& u, Vector& du_dt) const {
  // Compute:
  //    du_dt = M^{-1}*-K(u)
  // for du_dt
  Kmat.Mult(u, z);
  z.Neg();  // z = -z

  ParGridFunction tmp_u(&fespace), tmp_z(&fespace), tmp_du_dt(&fespace);
  tmp_u.SetFromTrueDofs(u);
  tmp_z.SetFromTrueDofs(z);
  tmp_du_dt.SetFromTrueDofs(du_dt);

  // // Construct and set inhomogeneous Neumann condition
  // ParLinearForm neumann(&fespace);
  // ConstantCoefficient ten(10.0);
  // Array<int> neumann_bdr(fespace.GetParMesh()->bdr_attributes.Max());
  // neumann_bdr = 0;
  // neumann_bdr[1] = 1;  // Right wall for inline-quad.mesh
  // neumann_bdr[3] = 1;  // Left wal for inline-quad.mesh
  // neumann.AddBoundaryIntegrator(new mfem::BoundaryLFIntegrator(ten),
  //                               neumann_bdr);
  // neumann.Assemble();

  // // Add inhomogeneous Neumann to RHS
  // tmp_z.Add(1.0, neumann);

  OperatorHandle A;
  Vector X, B;
  K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);

  M_solver.Mult(B, du_dt);
  M->RecoverFEMSolution(du_dt, tmp_z, tmp_du_dt);
  tmp_du_dt.GetTrueDofs(du_dt);
  du_dt.SetSubVector(ess_tdof_list, 0.0);
}
```
Even for just the Dirichlet case (as above), running in parallel leads to a different solution. Using 4 cores, invoked with 
` mpirun -np 4 example -visit -rs 0 -rp 0 -s 11 -dt 0.001 -o 1` , leads to the result below compared to the serial case 3 posts above.

![dirichlet_only_4](https://user-images.githubusercontent.com/47005129/95907661-1de92080-0d59-11eb-8595-3097df0cb634.png)

In case it is helpful, here is the same picture with the blocks (which I believe is the subdomain each element belongs to) labeled. 

![dirichlet_only_4_blocks](https://user-images.githubusercontent.com/47005129/95907768-4e30bf00-0d59-11eb-960a-80cc8cd209b7.png)

**Comment by mlstowell** [comment]
Hi, @robert-chiodi ,

I think I see the problem: `tmp_z` is declared as a `ParGridFunction` but it is used to store values from a dual vector (see [Primal and Dual Vectors](https://mfem.org/pri-dual-vec/).). This is a problem because `Kmat.Mult(u, z)` places fully summed linear form values into `z`.  The call to `tmp_z.SetFromTrueDofs(z);` then duplicates these shared values on all processor boundaries (which makes sense for `ParGridFunction`s but not `ParLinearForm`s).   Finally, the call to `K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);` sums the duplicated values from the processor boundaries to attempt to make fully summed linear form entries in `B` but of course this results in those values being doubled (or even quadrupled at the central node).

The fix is to use `K` rather than `Kmat` to perform the multiplication. 
```
ParGridFunction tmp_u(&fespace), tmp_du_dt(&fespace);
ParLinearForm tmp_z(&fespace);
tmp_u.SetFromTrueDofs(u);
tmp_du_dt.SetFromTrueDofs(du_dt);
K->Mult(tmp_u, tmp_z);

OperatorHandle A;
Vector X, B;
K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);

...
```

I was looking mainly at `ConductionOperator::Mult` but the same applies to the `ImplicitSolve` method.  Also, the `neumann` vector can be added to this `tmp_z` and it should work as expected.

I hope this helps,
Mark

**Comment by robert-chiodi** [comment]
Hi Mark,

Thank you very much for your response. Taking your comments into consideration, the `Mult` routine now looks like
```
void ConductionOperator::Mult(const Vector& u, Vector& du_dt) const {
  // Compute:
  //    du_dt = M^{-1}*-K(u)
  // for du_dt
  ParGridFunction tmp_u(&fespace), tmp_du_dt(&fespace);
  ParLinearForm tmp_z(&fespace);
  tmp_u.SetFromTrueDofs(u);
  tmp_du_dt.SetFromTrueDofs(du_dt);
  K->Mult(tmp_u, tmp_z);  // <----- Segfaults in Mult routine
  tmp_z.Neg();            // z = -z

  // OperatorHandle K_op;
  // K->FormSystemMatrix(ess_tdof_list, K_op);
  // K_op->Mult(tmp_u, tmp_z);  // <--- Works in serial, mismatch in
  //                            // matrix/vector sizes when parallel
  // tmp_z.Neg();               // z = -z

  // Construct and set inhomogeneous Neumann condition
  ParLinearForm neumann(&fespace);
  ConstantCoefficient ten(10.0);
  Array<int> neumann_bdr(fespace.GetParMesh()->bdr_attributes.Max());
  neumann_bdr = 0;
  neumann_bdr[1] = 1;  // Right wall for inline-quad.mesh
  neumann_bdr[3] = 1;  // Left wal for inline-quad.mesh
  neumann.AddBoundaryIntegrator(new mfem::BoundaryLFIntegrator(ten),
                                neumann_bdr);
  neumann.Assemble();

  // Add inhomogeneous Neumann to RHS
  tmp_z.Add(1.0, neumann);

  OperatorHandle A;
  Vector X, B;
  K->FormLinearSystem(ess_tdof_list, tmp_u, tmp_z, A, X, B);

  M_solver.Mult(B, du_dt);
  M->RecoverFEMSolution(du_dt, tmp_z, tmp_du_dt);
  tmp_du_dt.GetTrueDofs(du_dt);
  du_dt.SetSubVector(ess_tdof_list, 0.0);
}
```
Please let me know if I did not understand you correctly and this was not what you were suggesting. When I try to use this, in serial or parallel, there is a Segfault inside ` K->Mult(tmp_u, tmp_z);` due to an unmapped address. The stack trace is 
```
*** Process received signal ***
Signal: Segmentation fault: 11 (11)
Signal code: Address not mapped (1)
Failing at address: 0x0
[ 0] 0   libsystem_platform.dylib            0x00007fff6b243b5d _sigtramp + 29
[ 1] 0   libsystem_malloc.dylib              0x00007fff6b20420f szone_malloc_should_clear + 66
[ 2] 0   example                             0x0000000105a3cefc _ZNK18ConductionOperator4MultERKN4mfem6VectorERS1_ + 112
[ 3] 0   example                             0x0000000105a9a558 _ZN4mfem18ForwardEulerSolver4StepERNS_6VectorERdS3_ + 114
[ 4] 0   example                             0x0000000105a3aa6e main + 9309
[ 5] 0   libdyld.dylib                       0x00007fff6b0583d5 start + 1
[ 6] 0   ???                                 0x000000000000000c 0x0 + 12
*** End of error message ***
```
I checked, and `K` is indeed allocated and assembled prior to its use in `Mult`. I also tried forming the operator `K_op` first and using that for the multiplication, which works in serial but not in parallel (due to differences in matrix/vector sizes). I suspect use of `K_op` would lead to the same problem you were discussing before anyways though. 

I am hoping the SegFault in `K->Mult()` is just an error I have made. If you see any mistake I made trying to implement what you suggested, please let me know.

**Comment by chang-yiyi** [comment]
Hi @robert-chiodi ,

Maybe you can try comment `K->FormSystemMatrix(ess_tdof_list, Kmat)` in `void ConductionOperator::SetParameters(const Vector& u)`, like this
```
void ConductionOperator::SetParameters(const Vector& u) {
    ParGridFunction u_alpha_gf(&fespace);
    u_alpha_gf.SetFromTrueDofs(u);
    for (int i = 0; i < u_alpha_gf.Size(); i++) {
        u_alpha_gf(i) = kappa + alpha * u_alpha_gf(i);
    }

    delete K;
    K = new ParBilinearForm(&fespace);

    GridFunctionCoefficient u_coeff(&u_alpha_gf);

    K->AddDomainIntegrator(new DiffusionIntegrator(u_coeff));
    K->Assemble(0);  // keep sparsity pattern of M and K the same
//    K->FormSystemMatrix(ess_tdof_list, Kmat);
    delete T;
    T = NULL;  // re-compute T on the next ImplicitSolve
}

```

**Comment by robert-chiodi** [comment]
Hi @chang-yiyi ,

Thank you very much for the suggestion, that worked! I changed a few other things as well, and it seems to be working in serial and parallel now. For my own understanding, do you know in what way `FormSystemMatrix` modifies the `ParBilinearForm` `K`? I am surprised it is not just a factory method to create the `HypreParMatrix` `Kmat`.

For completeness, I have included below the working `Mult()` function for use with explicit time schemes.

```
void ConductionOperator::Mult(const Vector& u, Vector& du_dt) const {
  // Compute:
  //    du_dt = M^{-1}*-K(u)
  // for du_dt
  ParGridFunction tmp_u(&fespace);
  ParLinearForm tmp_z(&fespace);
  tmp_u.SetFromTrueDofs(u);
  K->Mult(tmp_u, tmp_z);
  tmp_z.Neg();  // z = -z

  // Construct and set inhomogeneous Neumann condition
  ParLinearForm neumann(&fespace);
  ConstantCoefficient ten(10.0);
  Array<int> neumann_bdr(fespace.GetParMesh()->bdr_attributes.Max());
  neumann_bdr = 0;
  neumann_bdr[1] = 1;  // Right wall for inline-quad.mesh
  neumann_bdr[3] = 1;  // Left wal for inline-quad.mesh
  neumann.AddBoundaryIntegrator(new mfem::BoundaryLFIntegrator(ten),
                                neumann_bdr);
  neumann.Assemble();

  // Add inhomogeneous Neumann to RHS
  tmp_z.Add(1.0, neumann);

  OperatorHandle A;
  Vector X, B;
  ParGridFunction tmp_du_dt(&fespace);
  tmp_du_dt = 0.0;
  M->FormLinearSystem(ess_tdof_list, tmp_du_dt, tmp_z, A, du_dt, B);
  M_solver.Mult(B, du_dt);
}
```

**Comment by mlstowell** [comment]
Hi, @robert-chiodi ,

I should have caught this but you also need to `Finalize` the `K` operator before using it like this.  For `BilinearForm`s you should:
```
K->Assemble();
K->Finalize();
```
Then it should be ready to use.  The `FormSystemMatrix` calls `Finalize` for you but it also allocates and fills another copy of the matrix which is unnecessary in this case.

Best wishes,
Mark

**Comment by robert-chiodi** [comment]
Hi @mlstowell ,

Thank you so much, that did the trick! I really appreciate your (and everyone else's) help!

**Comment by chang-yiyi** [comment]
Hi @robert-chiodi ,

Did you try to use time-dependent Dirichlet and Neumann boundary conditions?

**Comment by robert-chiodi** [comment]
Hi @chang-yiyi ,

Currently, I can change the BCs in time by resetting the values in `u` prior to a call to `ode_solver->Step()`. Technically, I believe this is incorrect for higher-order time integration schemes. 

I think something would need to be done like you discussed in https://github.com/mfem/mfem/issues/1720#issuecomment-687563198 . As for how to achieve that system, I am not sure. Maybe some call to `M->FormLinearSystem` with appropriate (non-zero) values for du_dt at the ess_tdof locations? It might be worthwhile filing a new issue to see if someone with more experience can help you.

**Comment by v-dobrev** [comment]
Hi @robert-chiodi and @chang-yiyi, I just posted some suggestions about the time-dependent Dirichlet BCs here: https://github.com/mfem/mfem/issues/1720#issuecomment-709505317.

For time-dependent Neumann BCs: since these are imposed weakly as a source in the r.h.s., the only difference with the time-independent case is that one has to re-assemble the `(Par)LinearForm` after updating the time of the Neumann b.c. coefficient with `SetTime`.

**Comment by chang-yiyi** [comment]
Hi @robert-chiodi and @v-dobrev ,

Thanks for your reply.

**Comment by fengerhan** [comment]
Hi @robert-chiodi ,
You have incorporate inhomogeneous Dirichlet and Neumann conditions into the heat equation code of example 16. Can you post your serial code ?

**Comment by robert-chiodi** [comment]
Hi @fengerhan,

I am sorry for the incredibly late reply. I have spent a little bit of time constructing a modified ex16p with Dirichlet and Neumann conditions. I hope it can still be of some help to you, and I would be happy to answer any questions that might arise. @bensworth, I believe you were interested too.

In the modified ex16p below, I have focused on solving two specific 1D problems from
```
   John D. Reinert, Anubhav Dwivedi, and Graham V. Candler.
   "Verification of a conjugate heat transfer tool with US3D". AIAA  SciTech 2019. 
```

In particular, `problem_number == 1` (chosen with `-pn 1`) solves the heat equation with constant diffusivity for an initially 300 K domain and a boundary condition on the right wall of 500 K. For `problem_number == 2` (chosen with `-pn 2`), the problem consists of an initially 300 K domain and an in-flux of heat on the left wall. The specific heat and conductivity are also non-linearly dependent on the temperature in this case. 

For both problems the analytical solution is known and the error between the computed solution at final time (chosen by flag `tf`) and the analytical solution is printed to the screen. These solutions are computed inside the C++ lambda `solution_lambda`. It is important to note that convergence is not expected at "low enough" errors because the analytical solution is a truncation of the infinite series solution. This could be fixed by initializing the temperature `u` as the truncated analytical solution at time `t = 0`, but I have chosen not to do so. If identifying correct convergence behavior becomes important, it would be easy to implement this by projecting the initial condition onto `u`, where the provided `solution_lambda` could be used to construct a `FunctionCoefficient`.

The 1D nature of the problem should not have influenced the code too much, with the exception of knowing the boundary tag values for the left and right wall when setting Dirichlet or Neumann conditions.



```
//                       Modified MFEM Example 16 - Parallel Version
//
// Compile with: make ex16p
//
// Sample runs:  mpirun -np 1 ex16p -pn 1 --line-out
//               mpirun -np 4 ex16p -pn 1
//               mpirun -np 1 ex16p -pn 2 
//               mpirun -np 4 ex16p -pn 2
//
// Description:  This example solves a time dependent nonlinear heat equation
//               problem of the form \rho c_p du/dt = C(u), with a (potentially)
//               non-linear diffusion operator C(u) = \nabla \cdot (\kappa) \nabla u
//               and non-linear specific heat c_p.
//
//               Two problems can be solved in this example, both on a 1D domain
//               spanning [0, 0.01].
//
//               These come from the paper
//               John D. Reinert, Anubhav Dwivedi, and Graham V. Candler.
//               "Verification of a conjugate heat transfer tool with US3D".
//               AIAA SciTech 2019.
//
//               Problem 1 is the verification case in Section B.1, which consists
//               of a Dirichlet boundary condition on the right boundary (x = 0.01)
//               of 500 and a homogeneous Neumann condition on the left boundary.
//
//               Problem 2 is the variable thermal conductivity case in Section B.3,
//               which involves a heat-flux (inhomogeneous Neumann) condition on the
//               left face with a value of 7.5 x 10^5. The specific heat and
//               thermal conductivity also vary linearly with temperature.
//
//               The example demonstrates the use of nonlinear operators (the
//               class ConductionOperator defining C(u)), as well as their
//               implicit time integration. Note that implementing the method
//               ConductionOperator::ImplicitSolve is the only requirement for
//               high-order implicit (SDIRK) time integration. Optional saving
//               with ADIOS2 (adios2.readthedocs.io) is also illustrated.
//
//               We recommend viewing examples 2, 9 and 10 before viewing this
//               example.

#include "mfem.hpp"
#include <fstream>
#include <iostream>

using namespace std;
using namespace mfem;

/** After spatial discretization, the conduction model can be written as:
 *
 *     du/dt = M^{-1}(-Ku)
 *
 *  where u is the vector representing the temperature, M is the mass matrix,
 *  and K is the diffusion operator. Depending on the problem, K may or may
 *  not be non-linear.
 *
 *  Class ConductionOperator represents the right-hand side of the above ODE.
 */
class ConductionOperator : public TimeDependentOperator
{
protected:
   ParFiniteElementSpace &fespace;
   Array<int> active_bdr;
   Array<int> ess_tdof_list;
   long global_tdof_size;

   ParBilinearForm *M;
   ParBilinearForm *K;

   HypreParMatrix* Full_Kmat;  
   HypreParMatrix* Mmat;
   HypreParMatrix* Kmat;
   HypreParMatrix* Tmat; // T = M + dt K  
   HypreParMatrix* Mmat_e;
   HypreParMatrix* Kmat_e;
   HypreParMatrix* Tmat_e;  
   double current_dt;

   CGSolver M_solver;    // Krylov solver for inverting the mass matrix M
   HypreSmoother M_prec; // Preconditioner for the mass matrix M

   CGSolver T_solver;    // Implicit solver for T = M + dt K
   HypreSmoother T_prec; // Preconditioner for the implicit solver

   mutable Vector z; // auxiliary vector

   Vector neumann_vector;
   ParLinearForm* neumann;

public:
   ConductionOperator(ParFiniteElementSpace &f, Vector &u);

   virtual void Mult(const Vector &u, Vector &du_dt) const;
   /** Solve the Backward-Euler equation: k = f(u + dt*k, t), for the unknown k.
       This is the only requirement for high-order SDIRK implicit integration.*/
   virtual void ImplicitSolve(const double dt, const Vector &u, Vector &k);

   /// Update the diffusion BilinearForm K using the given true-dof vector `u`.
   void SetParameters(const Vector &u);

   virtual ~ConductionOperator();
};

double InitialTemperature(const Vector &x);

int problem_number = 0;

int main(int argc, char *argv[])
{
   // 1. Initialize MPI.
   int num_procs, myid;
   MPI_Init(&argc, &argv);
   MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
   MPI_Comm_rank(MPI_COMM_WORLD, &myid);

   // 2. Parse command-line options.
   int nx = 10;
   int ser_ref_levels = 2;
   int par_ref_levels = 1;
   int order = 2;
   int ode_solver_type = 1;
   int problem = 0;
   double t_final = 4.0;
   double dt = 1.0e-2;
   bool visualization = true;
   bool visit = false;
   int vis_steps = 5;
   bool adios2 = false;
   bool line_out_data = false;

   int precision = 8;
   cout.precision(precision);

   OptionsParser args(argc, argv);
   args.AddOption(&nx, "-nx", "--nx","Number of elements for 1D domain");
   args.AddOption(&problem_number,"-pn","--problem-number","Problem to be solved:"
		  "0. constant conductivity with Dirichlet conditions"
		  "1. variable conductivity with Neumann conditions");
   args.AddOption(&ser_ref_levels, "-rs", "--refine-serial",
                  "Number of times to refine the mesh uniformly in serial.");
   args.AddOption(&par_ref_levels, "-rp", "--refine-parallel",
                  "Number of times to refine the mesh uniformly in parallel.");
   args.AddOption(&order, "-o", "--order",
                  "Order (degree) of the finite elements.");
   args.AddOption(&ode_solver_type, "-s", "--ode-solver",
                  "ODE solver: 1 - Backward Euler, 2 - SDIRK2, 3 - SDIRK3,\n\t"
                  "\t   11 - Forward Euler, 12 - RK2, 13 - RK3 SSP, 14 - RK4.");
   args.AddOption(&t_final, "-tf", "--t-final",
                  "Final time; start time is 0.");
   args.AddOption(&dt, "-dt", "--time-step",
                  "Time step.");
   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
                  "--no-visualization",
                  "Enable or disable GLVis visualization.");
   args.AddOption(&visit, "-visit", "--visit-datafiles", "-no-visit",
                  "--no-visit-datafiles",
                  "Save data files for VisIt (visit.llnl.gov) visualization.");
   args.AddOption(&vis_steps, "-vs", "--visualization-steps",
                  "Visualize every n-th timestep.");
   args.AddOption(&adios2, "-adios2", "--adios2-streams", "-no-adios2",
                  "--no-adios2-streams",
                  "Save data using adios2 streams.");
   args.AddOption(&line_out_data, "-line", "--line-out", "-no-line",
                  "--no-line-out",
                  "Will print a line out of the solution and the analytical solution."
		  " Only available in serial.");
   args.Parse();
   if (!args.Good())
   {
      args.PrintUsage(cout);
      MPI_Finalize();
      return 1;
   }

   if (myid == 0)
   {
      args.PrintOptions(cout);
   }

   MFEM_ASSERT(problem_number > 0 && problem_number < 3,
	       "Problem number must be either 1 or 2");
   MFEM_ASSERT(line_out_data  && num_procs == 1 || !line_out_data,
	       "Line out of result can only be produced in serial.");   

   // 3. Read the serial mesh from the given mesh file on all processors. We can
   //    handle triangular, quadrilateral, tetrahedral and hexahedral meshes
   //    with the same code.   
   Mesh *mesh = new Mesh(nx, 0.01);
   int dim = mesh->Dimension();
   MFEM_ASSERT(dim == 1,
	       "This example is designed to be used exclusively"
	       "with a 1D domain covering [0, 0.01]");

   // 4. Define the ODE solver used for time integration. Several implicit
   //    singly diagonal implicit Runge-Kutta (SDIRK) methods, as well as
   //    explicit Runge-Kutta methods are available.
   ODESolver *ode_solver;
   switch (ode_solver_type)
   {
      // Implicit L-stable methods
      case 1:  ode_solver = new BackwardEulerSolver; break;
      case 2:  ode_solver = new SDIRK23Solver(2); break;
      case 3:  ode_solver = new SDIRK33Solver; break;
      // Explicit methods
      case 11: ode_solver = new ForwardEulerSolver; break;
      case 12: ode_solver = new RK2Solver(0.5); break; // midpoint method
      case 13: ode_solver = new RK3SSPSolver; break;
      case 14: ode_solver = new RK4Solver; break;
      case 15: ode_solver = new GeneralizedAlphaSolver(0.5); break;
      // Implicit A-stable methods (not L-stable)
      case 22: ode_solver = new ImplicitMidpointSolver; break;
      case 23: ode_solver = new SDIRK23Solver; break;
      case 24: ode_solver = new SDIRK34Solver; break;
      default:
         cout << "Unknown ODE solver type: " << ode_solver_type << '\n';
         delete mesh;
         return 3;
   }

   // 5. Refine the mesh in serial to increase the resolution. In this example
   //    we do 'ser_ref_levels' of uniform refinement, where 'ser_ref_levels' is
   //    a command-line parameter.
   for (int lev = 0; lev < ser_ref_levels; lev++)
   {
      mesh->UniformRefinement();
   }

   // 6. Define a parallel mesh by a partitioning of the serial mesh. Refine
   //    this mesh further in parallel to increase the resolution. Once the
   //    parallel mesh is defined, the serial mesh can be deleted.
   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
   delete mesh;
   for (int lev = 0; lev < par_ref_levels; lev++)
   {
      pmesh->UniformRefinement();
   }

   // 7. Define the vector finite element space representing the current and the
   //    initial temperature, u_ref.
   H1_FECollection fe_coll(order, dim);
   ParFiniteElementSpace fespace(pmesh, &fe_coll);

   HYPRE_BigInt fe_size = fespace.GlobalTrueVSize();
   if (myid == 0)
   {
      cout << "Number of temperature unknowns: " << fe_size << endl;
   }

   ParGridFunction u_gf(&fespace);

   // 8. Set the initial conditions for u. All boundaries are considered
   //    natural.
   FunctionCoefficient u_0(InitialTemperature);
   u_gf.ProjectCoefficient(u_0);
   Vector u;
   u_gf.GetTrueDofs(u);

   // 9. Initialize the conduction operator and the VisIt visualization.
   ConductionOperator oper(fespace, u);
   u_gf.SetFromTrueDofs(u); 
   {
      ostringstream mesh_name, sol_name;
      mesh_name << "ex16-mesh." << setfill('0') << setw(6) << myid;
      sol_name << "ex16-init." << setfill('0') << setw(6) << myid;
      ofstream omesh(mesh_name.str().c_str());
      omesh.precision(precision);
      pmesh->Print(omesh);
      ofstream osol(sol_name.str().c_str());
      osol.precision(precision);
      u_gf.Save(osol);
   }

   VisItDataCollection visit_dc("Example16-Parallel", pmesh);
   visit_dc.RegisterField("temperature", &u_gf);
   if (visit)
   {
      visit_dc.SetCycle(0);
      visit_dc.SetTime(0.0);
      visit_dc.Save();
   }

   // Optionally output a BP (binary pack) file using ADIOS2. This can be
   // visualized with the ParaView VTX reader.
#ifdef MFEM_USE_ADIOS2
   ADIOS2DataCollection* adios2_dc = NULL;
   if (adios2)
   {
      const std::string collection_name = "ex16-p-1D.bp";

      adios2_dc = new ADIOS2DataCollection(MPI_COMM_WORLD, collection_name, pmesh);
      adios2_dc->SetParameter("SubStreams", std::to_string(num_procs/2) );
      adios2_dc->RegisterField("temperature", &u_gf);
      adios2_dc->SetCycle(0);
      adios2_dc->SetTime(0.0);
      adios2_dc->Save();
   }
#endif

   socketstream sout;
   if (visualization)
   {
      char vishost[] = "localhost";
      int  visport   = 19916;
      sout.open(vishost, visport);
      sout << "parallel " << num_procs << " " << myid << endl;
      int good = sout.good(), all_good;
      MPI_Allreduce(&good, &all_good, 1, MPI_INT, MPI_MIN, pmesh->GetComm());
      if (!all_good)
      {
         sout.close();
         visualization = false;
         if (myid == 0)
         {
            cout << "Unable to connect to GLVis server at "
                 << vishost << ':' << visport << endl;
            cout << "GLVis visualization disabled.\n";
         }
      }
      else
      {
         sout.precision(precision);
         sout << "solution\n" << *pmesh << u_gf;
         sout << "pause\n";
         sout << flush;
         if (myid == 0)
         {
            cout << "GLVis visualization paused."
                 << " Press space (in the GLVis window) to resume it.\n";
         }
      }
   }   

   // 10. Perform time-integration (looping over the time iterations, ti, with a
   //     time-step dt).
   ode_solver->Init(oper);
   double t = 0.0;

   bool last_step = false;
   for (int ti = 1; !last_step; ti++)
   {
      if (t + dt >= t_final - dt/2)
      {
         last_step = true;
      }

      ode_solver->Step(u, t, dt);

      if (last_step || (ti % vis_steps) == 0)
      {
         if (myid == 0)
         {
            cout << "step " << ti << ", t = " << t << endl;
         }

         u_gf.SetFromTrueDofs(u);
         if (visualization)
         {
            sout << "parallel " << num_procs << " " << myid << "\n";
            sout << "solution\n" << *pmesh << u_gf << flush;
         }

         if (visit)
         {
            visit_dc.SetCycle(ti);
            visit_dc.SetTime(t);
            visit_dc.Save();
         }

#ifdef MFEM_USE_ADIOS2
         if (adios2)
         {
            adios2_dc->SetCycle(ti);
            adios2_dc->SetTime(t);
            adios2_dc->Save();
         }
#endif
      }
      oper.SetParameters(u);
   }

#ifdef MFEM_USE_ADIOS2
   if (adios2)
   {
      delete adios2_dc;
   }
#endif

   
   // 11. Save the final solution in parallel. This output can be viewed later
   //     using GLVis: "glvis -np <np> -m ex16-mesh -g ex16-final".
   {
      ostringstream sol_name;
      sol_name << "ex16-final." << setfill('0') << setw(6) << myid;
      ofstream osol(sol_name.str().c_str());
      osol.precision(precision);
      u_gf.Save(osol);
   }

   // Compute error
   if(problem_number == 1){

     static constexpr double diffusivity = 2.5e-6;
     static constexpr int approximation_terms = 10;
     static constexpr double L = 0.01;
     
     auto solution_lambda = [=](const double* a_position, const double a_time) {
       double approx = 0.0;
       for (int n = 0; n < approximation_terms; ++n) {
	 const double dn = static_cast<double>(n);
	 approx += std::pow(-1.0, n) / (2.0 * dn + 1.0) *
	   std::exp(-0.25 * std::pow((2.0 * dn + 1.0) * M_PI, 2.0) *
		    diffusivity * a_time / (L * L)) *
	   std::cos(0.5 * (2.0 * dn + 1.0) * M_PI * a_position[0] / L);
       }
       return 300 + (500 - 300) * (1.0 - 4.0 / M_PI * approx);
     };
     auto solution_at_time = [t,solution_lambda](const Vector& x){
       return solution_lambda(x, t);
     };
     
     FunctionCoefficient sol_coef(solution_at_time);
     double l1_value = u_gf.ComputeL1Error(sol_coef);
     double l2_value = u_gf.ComputeL2Error(sol_coef);
     double linf_value = u_gf.ComputeMaxError(sol_coef);
     if(myid == 0){
       std::cout << "Errors are \n";
       std::cout << "L1 Error: " << l1_value << '\n';
       std::cout << "L2 Error: " << l2_value << '\n';
       std::cout << "Linf Error: " << linf_value << std::endl;

       // Plot line-out of data 
       // Note: Doing this only makes sense in serial
       if(num_procs == 1 && line_out_data){
	 Vector x_loc(1);
	 IntegrationPoint ip;
	 std::cout << "Time    Computed T     Correct T" << std::endl;	 
	 for(int n = 0 ; n < pmesh->GetNE(); ++n){
	   ElementTransformation* trans = pmesh->GetElementTransformation(n);
	   ip.Set1w(0.0, 1.0);
	   trans->Transform(ip, x_loc);	 
	   std::cout << x_loc(0) << " " << u_gf.GetValue(*trans, ip, 1) << " "
		     << solution_at_time(x_loc) << std::endl;
	   ip.Set1w(0.5, 1.0);
	   trans->Transform(ip, x_loc);
	   std::cout << x_loc(0) << " " << u_gf.GetValue(*trans, ip, 1) << " "
		     << solution_at_time(x_loc) << std::endl;	 
	 }
       }
     }
   }

   if(problem_number == 2){

     static constexpr int approximation_terms = 10;
     static constexpr double L = 0.01;
     static constexpr double flux = 7.5e5;
     static constexpr double k_1 = 10.0;
     static constexpr double k_2 = 100.0;
     static constexpr double T_0 = 300.0;
     static constexpr double T_1 = 300.0;
     static constexpr double T_2 = 1300.0;
     static constexpr double C_1 = 500.0;
     //  static constexpr double C_2 = 5000.0; // Not needed in solution
     static constexpr double diffusivity = k_1 / (8000.0 * C_1);
     static constexpr double slope = (k_2 - k_1) / (T_2 - T_1);

     auto solution_lambda = [=](const double* a_position, const double a_time) {
       double approx = 0.0;
       for (int n = 1; n <= approximation_terms; ++n) {
	 const double dn = static_cast<double>(n);
	 approx +=
	   1.0 / std::pow(dn, 2.0) *
	   std::exp(-dn * dn * M_PI * M_PI * diffusivity * a_time / (L * L)) *
	   std::cos(dn * M_PI * a_position[0] / L);
       }
       const double theta_0 =
	 (T_0 - T_1) + slope / (2.0 * k_1) * (T_0 - T_1) * (T_0 - T_1);
       const double theta =
	 theta_0 +
	 flux * L / k_1 *
	 (diffusivity * a_time / (L * L) + 1.0 / 3.0 - a_position[0] / L +
	  0.5 * (a_position[0] * a_position[0] / (L * L)) -
	  2.0 / (M_PI * M_PI) * approx);
       return (std::sqrt(2.0 * k_1) * std::sqrt(2.0 * k_1 + 4.0 * theta * slope) -
	       2.0 * k_1 + 2.0 * T_1 * slope) /
	 (2.0 * slope);
     };
     auto solution_at_time = [t,solution_lambda](const Vector& x){
       return solution_lambda(x, t);
     };     
     
     FunctionCoefficient sol_coef(solution_at_time);
     double l1_value = u_gf.ComputeL1Error(sol_coef);
     double l2_value = u_gf.ComputeL2Error(sol_coef);
     double linf_value = u_gf.ComputeMaxError(sol_coef);
     if(myid == 0){
       std::cout << "Errors are \n";
       std::cout << "L1 Error: " << l1_value << '\n';
       std::cout << "L2 Error: " << l2_value << '\n';
       std::cout << "Linf Error: " << linf_value << std::endl;

       // Plot line-out of data 
       // Note: Doing this only makes sense in serial
       if(num_procs == 1 && line_out_data){
	 Vector x_loc(1);
	 IntegrationPoint ip;
	 std::cout << "Time    Computed T     Correct T" << std::endl;
	 for(int n = 0 ; n < pmesh->GetNE(); ++n){
	   ElementTransformation* trans = pmesh->GetElementTransformation(n);
	   ip.Set1w(0.0, 1.0);
	   trans->Transform(ip, x_loc);	 
	   std::cout << x_loc(0) << " " << u_gf.GetValue(*trans, ip, 1) << " "
		     << solution_at_time(x_loc) << std::endl;
	   ip.Set1w(0.5, 1.0);
	   trans->Transform(ip, x_loc);
	   std::cout << x_loc(0) << " " << u_gf.GetValue(*trans, ip, 1) << " "
		     << solution_at_time(x_loc) << std::endl;	 
	 }
       }
     }
   }

   // 12. Free the used memory.
   delete ode_solver;
   delete pmesh;

   MPI_Finalize();

   return 0;
}

ConductionOperator::ConductionOperator(ParFiniteElementSpace &f, Vector &u)
  : TimeDependentOperator(f.GetTrueVSize(), 0.0), fespace(f), active_bdr(),
    ess_tdof_list(), global_tdof_size(0), M(NULL), K(NULL),
    Mmat(NULL), Full_Kmat(NULL), Kmat(NULL), Tmat(NULL),
    Mmat_e(NULL), Kmat_e(NULL), Tmat_e(NULL), current_dt(0.0),
    M_solver(f.GetComm()), T_solver(f.GetComm()), z(height),
    neumann(nullptr), neumann_vector()
{
   const double rel_tol = 1e-8;

   // Problem 1 has Dirichlet boundary condition on right side
   // Problem 2 has Neumann condition on left side
   active_bdr.SetSize(f.GetParMesh()->bdr_attributes.Max());
   active_bdr = 0;
   active_bdr[problem_number == 1 ? 1 : 0] = 1;

   // Set up solvers since needed in either case
   M_solver.iterative_mode = false;
   M_solver.SetRelTol(rel_tol);
   M_solver.SetAbsTol(0.0);
   M_solver.SetMaxIter(500);
   M_solver.SetPrintLevel(0);
   M_prec.SetType(HypreSmoother::Jacobi);
   M_solver.SetPreconditioner(M_prec);

     
   T_solver.iterative_mode = false;
   T_solver.SetRelTol(rel_tol);
   T_solver.SetAbsTol(0.0);
   T_solver.SetMaxIter(500);
   T_solver.SetPrintLevel(0);
   T_solver.SetPreconditioner(T_prec);       
   
   if(problem_number == 1){

     // Get Essential True Dofs 
     fespace.GetEssentialTrueDofs(active_bdr, ess_tdof_list);
     global_tdof_size = f.GetParMesh()->ReduceInt(ess_tdof_list.Size());
     
     // Set value at essential DOFs
     static constexpr double T_border = 500.0;
     u.SetSubVector(ess_tdof_list, T_border);
   
     M = new ParBilinearForm(&fespace);
     static constexpr double rho = 8000.0;
     static constexpr double c_p = 500.0;
     ConstantCoefficient rho_c_p_coeff(rho*c_p);
     
     M->AddDomainIntegrator(new MassIntegrator(rho_c_p_coeff));
     M->Assemble(0); // keep sparsity pattern of M and K the same
     M->Finalize(0);     
     Mmat = M->ParallelAssemble();
     Mmat_e = M->ParallelEliminateTDofs(ess_tdof_list, *Mmat);   
     
     M_solver.SetOperator(*Mmat); 

     K = new ParBilinearForm(&fespace);    
     ConstantCoefficient conductivity_coeff(10.0); 
     
     K->AddDomainIntegrator(new DiffusionIntegrator(conductivity_coeff));
     K->Assemble(0); // keep sparsity pattern of M and K the same
     K->Finalize(0); 
     Kmat = K->ParallelAssemble();
     Full_Kmat = new HypreParMatrix(*Kmat);
     Kmat_e = K->ParallelEliminateTDofs(ess_tdof_list, *Kmat);    
   }

   if(problem_number == 2){
     // Generate Neumann condition
     // Remains constant as flux value remains the same
     // and mesh is not changed.
     neumann = new ParLinearForm(&fespace);
     static constexpr double flux_value = 7.5e5;
     ConstantCoefficient flux_coeff(flux_value);
     neumann->AddBoundaryIntegrator(new BoundaryLFIntegrator(flux_coeff), active_bdr);
     neumann->Assemble();
     neumann_vector.SetSize(fespace.TrueVSize());
     neumann->ParallelAssemble(neumann_vector);     
   }

   SetParameters(u);
}

void ConductionOperator::Mult(const Vector &u, Vector &du_dt) const
{
   // Compute:
   //    du_dt = M^{-1}*-K(u)
   // for du_dt
   Full_Kmat->Mult(u, z);
   z.Neg(); // z = -z

   if(problem_number == 2){
     // Add Neumann condition
     z.Add(1.0, neumann_vector);
   }

   if(global_tdof_size > 0){   
     du_dt.SetSubVector(ess_tdof_list, 0.0);   
     EliminateBC(*Mmat,*Mmat_e, ess_tdof_list, du_dt, z);
   }
   M_solver.Mult(z, du_dt);
}

void ConductionOperator::ImplicitSolve(const double dt,
                                       const Vector &u, Vector &du_dt)
{
   // Solve the equation:
   //    du_dt = M^{-1}*[-K(u + dt*du_dt)]
   // for du_dt
  
  // Forces rebuild of T for problem_number == 1 when dt changes
  if(current_dt != dt && problem_number == 1){
    delete Tmat;
    delete Tmat_e;
    Tmat = nullptr;
    Tmat_e = nullptr;
  }
  if (!Tmat)
   {
      Tmat = Add(1.0, *Mmat, dt, *Kmat);
      if(global_tdof_size > 0){	
	Tmat_e = Add(1.0, *Mmat_e, dt, *Kmat_e);
      }
      current_dt = dt;
      T_solver.SetOperator(*Tmat);
   }
   MFEM_VERIFY(dt == current_dt, ""); // SDIRK methods use the same dt
   Full_Kmat->Mult(u, z);
   z.Neg();

   if(problem_number == 2){
     // Add Neumann condition
     z.Add(1.0, neumann_vector);
   }

   if(global_tdof_size > 0){
     du_dt.SetSubVector(ess_tdof_list, 0.0);
     EliminateBC(*Tmat,*Tmat_e, ess_tdof_list, du_dt, z);
   }
   T_solver.Mult(z, du_dt);
}

void ConductionOperator::SetParameters(const Vector &u)
{

  // Problem 1 has invariant rho, c_p, and conductivity
  // so rebuilding is not necessary

  // Problem 2 has changing conductivity and specific heat
  //
  // Note, we do not actually need to Eliminate BCs since
  // Problem 2 has no essential DOFs. We have kept the
  // procedure to enable easier changing to problems including
  // Dirichlet conditions.

  if(problem_number == 2){
    ParGridFunction conductivity(&fespace);
    ParGridFunction rho_cp(&fespace);    
    conductivity.SetFromTrueDofs(u);
    rho_cp = conductivity;
    
    static constexpr double k1 = 10.0;
    static constexpr double k2 = 100.0;
    static constexpr double T1 = 300.0;
    static constexpr double T2 = 1300.0;
    static constexpr double C1 = 500.0;
    static constexpr double C2 = 5000.0;         
    for (int i = 0; i < conductivity.Size(); i++)
    {
      conductivity(i) = k1 + (k2-k1)/(T2-T1)*(conductivity(i)-T1);
    }
    static constexpr double rho = 8000.0;
    for (int i = 0; i < rho_cp.Size(); i++)
    {
      rho_cp(i) = rho*(C1 + (C2-C1)/(T2-T1)*(rho_cp(i)-T1));
    }
    
    delete K;
    if(global_tdof_size > 0){      
      delete Full_Kmat; // Otherwise, Full_Kmat just points to Kmat.
    }
    delete Kmat;
    delete Kmat_e;
    delete M;
    delete Mmat;
    delete Mmat_e;
    delete Tmat;
    delete Tmat_e;
    Tmat = nullptr; // Will force rebuilding in ImplicitSolve

    M = new ParBilinearForm(&fespace);
    GridFunctionCoefficient rho_c_p_coeff(&rho_cp);

    M->AddDomainIntegrator(new MassIntegrator(rho_c_p_coeff));
    M->Assemble(0); // keep sparsity pattern of M and K the same
    M->Finalize(0);     
    Mmat = M->ParallelAssemble();
    if(global_tdof_size > 0){
      Mmat_e = M->ParallelEliminateTDofs(ess_tdof_list, *Mmat);
    }
         
    M_solver.SetOperator(*Mmat);       

    K = new ParBilinearForm(&fespace);    
    GridFunctionCoefficient conductivity_coeff(&conductivity);
    K->AddDomainIntegrator(new DiffusionIntegrator(conductivity_coeff));
    K->Assemble(0); // keep sparsity pattern of M and K the same
    K->Finalize(0); 
    Kmat = K->ParallelAssemble();
    if(global_tdof_size > 0){    
      Full_Kmat = new HypreParMatrix(*Kmat);
      Kmat_e = K->ParallelEliminateTDofs(ess_tdof_list, *Kmat);
    } else {
      Full_Kmat = Kmat;
    }
  }
}

ConductionOperator::~ConductionOperator()
{
  delete M;
  delete K;
  if(global_tdof_size > 0){
    delete Full_Kmat;
  }
  delete Mmat;
  delete Kmat;
  delete Tmat;
  delete Mmat_e;
  delete Kmat_e;
  delete Tmat_e;
  delete neumann;
}

double InitialTemperature(const Vector &x)
{
  return 300.0;
}
```
